internal/client.go

```diff
@@ -16,152 +16,207 @@ import (
+	return func(c *Client) error {
+		c.cookies = cookies
+		return nil
+	}
+}
 
-		ctx := context.TODO()
-		req, err := baseClient.NewRequest(ctx, http.MethodGet, "/users/me", nil)
-		if err != nil {
-			return baseClient, err
+// WithAbsolutePath sets the absolute path for the httpClient to communicate with
+func WithAbsolutePath(absolutePath string) ClientOption {
+	return func(c *Client) error {
+		c.absolutePath = absolutePath
+		return nil
+	}
+}
+
+// NewClient returns a wrapper around http/https (as per isHTTP flag) httpClient with additions of proxy & session_auth if given
+func NewClient(opts ...ClientOption) (*Client, error) {
+	c := &Client{
+		httpClient: http.DefaultClient,
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
 		}
+	}
 
-		resp, err := baseClient.client.Do(req)
+	// If the user does not specify a logger, then we'll use zap for a default one
+	if c.logger == nil {
+		zapLog, err := zap.NewProduction()
 		if err != nil {
-			return baseClient, err
+			return nil, err
 		}
-		defer func() {
-			if rerr := resp.Body.Close(); err == nil {
-				err = rerr
-			}
-		}()
-
-		err = CheckResponse(resp)
-
-		baseClient.Cookies = resp.Cookies()
+		logger := zapr.NewLogger(zapLog)
+		c.logger = &logger
 	}
 
-	return baseClient, nil
-}
-
-// NewBaseClient returns a basic http/https client based on isHttp flag
-func NewBaseClient(credentials *prismgoclient.Credentials, absolutePath string, isHTTP bool) (*Client, error) {
-	if absolutePath == "" {
+	if c.UserAgent == "" {
+		return nil, fmt.Errorf("userAgent argument must be passed")
+	}
+	if c.absolutePath == "" {
 		return nil, fmt.Errorf("absolutePath argument must be passed")
 	}
-
-	httpClient := http.DefaultClient
+	if c.credentials == nil {
+		return nil, fmt.Errorf("credentials argument must be passed")
+	}
+	if c.BaseURL == nil {
+		c.logger.V(1).Info("BaseURL is not set. Using URL from credentials", "url", c.credentials.URL)
+		if err := WithBaseURL(c.credentials.URL)(c); err != nil {
+			return nil, err
+		}
+	}
 
 	transCfg := &http.Transport{
 		// to skip/unskip SSL certificate validation
 		TLSClientConfig: &tls.Config{
-			InsecureSkipVerify: credentials.Insecure,
+			InsecureSkipVerify: c.credentials.Insecure,
 		},
 	}
-	httpClient.Transport = logging.NewTransport("Nutanix", transCfg)
-
-	protocol := httpsPrefix
-	if isHTTP {
-		protocol = httpPrefix
+	if c.credentials.ProxyURL != "" {
+		c.logger.V(1).Info("Using proxy:", "proxy", c.credentials.ProxyURL)
+		proxy, err := url.Parse(c.credentials.ProxyURL)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing proxy url: %s", err)
+		}
+		transCfg.Proxy = http.ProxyURL(proxy)
 	}
+	c.httpClient.Transport = transCfg
 
-	baseURL, err := url.Parse(fmt.Sprintf(defaultBaseURL, protocol, credentials.URL))
-	if err != nil {
-		return nil, err
-	}
+	if c.credentials.SessionAuth {
+		c.logger.V(1).Info("Using session_auth")
+
+		ctx := context.TODO()
```

Comment:
Can we at least create new context with some very high timeout like 1 minute? I'm thinking of cases where server stopped responding but is still reachable via network.

Reply:
Tbh, I don't like the sessionauth logic being present in the client constructor. I'd rather split it out into its own method. We inherited this client from the terraform repo and this PR is one of the many refactorings that we'll have to do to whip it into shape.